lib/datadog/security/reactive/engine.rb

```diff
@@ -0,0 +1,41 @@
+require 'datadog/security/reactive/address_hash'
+require 'datadog/security/reactive/subscriber'
+
+module Datadog
+  module Security
+    module Reactive
+      class Engine
+        def initialize
+          @data = {}
+          @subscribers = AddressHash.new { |h, k| h[k] = [] }
+          @children = []
+        end
+
+        def subscribe(*addresses, &block)
+          @subscribers[addresses.freeze] << Subscriber.new(&block).freeze
```

Comment:
Minor: I really like `freezing` everything we can! Have you considered moving the `freeze` to the `Subscriber` constructor directly? That way callers don't even need to remember to do that :)

Reply:
Noted